docs/cpu_templates/cpu-template-helper.md

```diff
@@ -0,0 +1,239 @@
+# CPU template helper tool
+
+The `cpu-template-helper` tool is a program designed to assist users with
+creating and managing their custom CPU templates.
+
+## Usage
+
+The `cpu-template-helper` tool has two sets of commands: template-related
+commands and fingerprint-related commands.
+
+### Template-related commands
+
+#### Dump command
+
+This command dumps guest CPU configuration in the custom CPU template JSON
+format.
+
+```
+cpu-template-helper template dump --config <firecracker-config> --output <cpu-config>
+```
+
+Users can utilize this as an entry point of a custom CPU template creation
+to comprehend what CPU configuration are exposed to guests.
+
+The guest CPU configuration consists of the following entities:
+
+- x86_64
+  - CPUID
+  - MSRs (Model Specific Registers)
+- aarch64
+  - ARM registers
+
+It retrieves the above entities exposed to a guest by applying the same preboot
+process as Firecacker and capturing them in the state just before booting a
+guest. More details about the preboot process can be found
+[here](boot-protocol.md) and [here](cpuid-normalization.md).
+
+> **Note**
+Some MSRs and ARM registers are not included in the output, since they are not
+reasonable to modify with CPU templates. The full list of them can be found in
+[Appendix](#appendix).
+
+> **Note**
+Since the output depends on underlying hardware and software stack (BIOS, CPU,
+kernel, Firecracker), it is required to dump guest CPU configuration on each
+combination when creating a custom CPU template targetting them all.
+
+#### Strip command
+
+This command strips identical entries from multiple guest CPU configuration
+files generated with the dump command.
+
+```
+cpu-template-helper template strip \
+    --paths <cpu-config-1> <cpu-config-2> [..<cpu-config-N>] \
+    --suffix <suffix>
+```
+
+One practical use case of the CPU template feature is to provide a consistent
+CPU feature set to guests running on multiple CPU models. When creating a custom
+CPU template for this purpose, it is efficient to focus on the differences
+in guest CPU configurations across those CPU models. Given that a dumped guest
+CPU configuration typically amounts to approximately 1,000 lines, this command
+considerably narrows down the scope to consider.
+
+#### Verify command
+
+This command verifies that the given custom CPU template is applied correctly.
+
+```
+cpu-template-helper template verify \
+    --config <firecracker-config>
+```
+
+Firecracker modifies the guest CPU configuration after the CPU template is
```

Comment:
We emphasize "after" here because the guest CPU config modification (CPUID normalization, boot protocol setting for MSRs and ARM registers) supersedes CPU template application. We also have the same description as cpuid-normalization.md.
https://github.com/firecracker-microvm/firecracker/blob/main/docs/cpu_templates/cpuid-normalization.md?plain=1#L6